src/contexts/ui/LoadingOverlay.tsx

```diff
@@ -0,0 +1,112 @@
+import PropTypes, { InferProps } from 'prop-types';
+import React, {
+  createContext,
+  useCallback,
+  useContext,
+  useEffect,
+  useMemo,
+  useRef,
+  useState,
+} from 'react';
+
+import uniqid from 'uniqid';
+import { Backdrop, CircularProgress } from '@material-ui/core';
+
+import {
+  LoadingOverlayConfig,
+  LoadingOverlayControls,
+  LoadingOverlayCtx,
+  LoadingOverlayInitializer,
+} from '@app/types';
+
+const LoadingOverlayContext = createContext<LoadingOverlayCtx>(null);
+
+const LoadingOverlayProvider: React.FC = props => {
+  const [registeredLoaders, setRegisteredLoaders] = useState<LoadingOverlayConfig[]>([]);
+  const isOpen: boolean = registeredLoaders.some(loader => loader.active);
+
+  const setLoaderStatus = useCallback((key: string, value: boolean) => {
+    setRegisteredLoaders((loaders: LoadingOverlayConfig[]) => {
+      const dupLoaders = [...loaders];
+      const requestedLoader = dupLoaders.find((loader: LoadingOverlayConfig) => {
+        return loader.key === key;
+      });
+      if (requestedLoader) {
+        requestedLoader.active = value;
+      }
+      return dupLoaders;
+    });
+  }, []);
+
+  const deregisterLoader = useCallback((key: string) => {
+    setRegisteredLoaders((loaders: LoadingOverlayConfig[]) => {
+      const dupLoaders = [...loaders];
+      const requestedLoaderIndex: number = dupLoaders.findIndex((loader: LoadingOverlayConfig) => {
+        return loader.key === key;
+      });
+      if (requestedLoaderIndex >= 0) {
+        dupLoaders.splice(requestedLoaderIndex, 1);
+      }
+      return dupLoaders;
+    });
+  }, []);
+
+  const registerLoader = useCallback((): LoadingOverlayControls => {
+    const key: string = uniqid();
+    setRegisteredLoaders(prevState => [...prevState, { key, active: false }]);
+    return {
+      activate: (): void => setLoaderStatus(key, true),
+      deactivate: (): void => setLoaderStatus(key, false),
+      deregister: (): void => deregisterLoader(key),
+    };
+  }, [deregisterLoader, setLoaderStatus]);
+
+  const contextValue: LoadingOverlayInitializer = useMemo(
+    () => ({
+      register: registerLoader,
+    }),
+    [registerLoader],
+  );
+
+  return (
+    <LoadingOverlayContext.Provider value={contextValue}>
+      <Backdrop style={{ zIndex: 1400 }} open={isOpen}>
+        {isOpen && <CircularProgress thickness={4} color="primary" />}
+      </Backdrop>
+      {props.children}
+    </LoadingOverlayContext.Provider>
+  );
+};
+
+const LoadingOverlayProps = {
+  open: PropTypes.bool,
+};
+
+const LoadingOverlay: React.FC<InferProps<typeof LoadingOverlayProps>> = props => {
```

Comment:
using this in our jsx composition feels a bit off to me; imo this functionality is best used as a hook - check out ffc576dd75255db1cc2eed65938ebf94a71a9c78

Reply:
That has problems once the loading overlay starts getting used in multiple places. We had a similar implementation at work and we had to change it to something like this. You have to attribute active and inactive loading states to specific consumers so they don't step on each other's toes. For example, say componentA opens the overlay and then componentB also opens the overlay a bit later. Now componentA's work is done and it closes the overlay but componentB would still like it to be open. 

This functionality could be achieved with a hook but binding it to a react component gives the advantage of allowing us to hook into the consumer's lifecycle and remove their loading state when they unmount. That way the consumer doesn't have to remember to close any open overlays they have when they unmount themselves.

Reply:
If you think this will be better in the long run then I'm not opposed. Only other note, can we drop proptypes here.
```suggestion
interface LoadingOverlayProps {
  open: boolean;
}
const LoadingOverlay: React.FC<LoadingOverlayProps> = props => {
```